server/src/main/java/org/opensearch/search/aggregations/bucket/terms/InternalSignificantTerms.java

```diff
@@ -265,13 +280,18 @@ public InternalAggregation reduce(List<InternalAggregation> aggregations, Reduce
             }
         }
         SignificanceHeuristic heuristic = getSignificanceHeuristic().rewrite(reduceContext);
-        final int size = reduceContext.isFinalReduce() == false ? buckets.size() : Math.min(requiredSize, buckets.size());
+        boolean isPartialReduce = reduceContext.isFinalReduce() == false && reduceContext.isSliceLevel() == false;
+        final int size = !isPartialReduce
+            ? Math.min(reduceContext.getRequiredSizeLocal(bucketCountThresholds), buckets.size())
+            : buckets.size();
         BucketSignificancePriorityQueue<B> ordered = new BucketSignificancePriorityQueue<>(size);
         for (Map.Entry<String, List<B>> entry : buckets.entrySet()) {
             List<B> sameTermBuckets = entry.getValue();
             final B b = reduceBucket(sameTermBuckets, reduceContext);
             b.updateScore(heuristic);
-            if (((b.score > 0) && (b.subsetDf >= minDocCount)) || reduceContext.isFinalReduce() == false) {
+            boolean meetsThresholds = (b.subsetDf >= reduceContext.getMinDocCountLocal(bucketCountThresholds))
+                && (((b.score > 0) || reduceContext.isSliceLevel()));
+            if (isPartialReduce || meetsThresholds) {
```

Comment:
can you please add comments explaining why this handling is needed here (like `for concurrent case we are not performing the shard level trimming in buildAggregation anymore and will be using reduce phase instead because of multiple slices......`)